src/Servers/Kestrel/Core/src/Internal/ConnectionDispatcher.cs

```diff
@@ -50,14 +51,21 @@ async Task AcceptConnectionsAsync()
 
                     // Add the connection to the connection manager before we queue it for execution
                     var id = _transportConnectionManager.GetNewConnectionId();
+
+                    // Cache counter enabled state at the start of the connection.
+                    // This ensures that the state is consistent for the entire connection.
+                    var metricsContext = new ConnectionMetricsContext(connection,
```

Comment:
I'm a little nervous about having the same properties (e.g. `CurrentConnectionsCounterEnabled`) on both the metrics object and the context - seems to easy to read the wrong one.  I wonder if it would be more maintainable to add a `MakeContext` method to `KestrelMetrics` that snapshots the appropriate flags and hands back an object?

Reply:
Good point. I found a bug like you describe in the PR to add metrics in rate limiting. It's not obvious at a glance that reading `_connectionDuration.Enabled` directly instead of from the context is bad but reading `_tlsHandshakeDuration.Enabled` directly is fine, so it's best keep all the code that decides whether to look at a cached or current enabled flag all in one place.

I don't see any bugs like that in this PR, but it's not easy to review. Making the "context" the one place singularly responsible for all the enabled checks even if it's just redirecting the call right back to `KestrelMetrics` is one option. I understand wanting to avoid indirection through an extra object though. Particularly, in the disabled case.

Another option might be a naming convention like `_connectionDurationUncached.Enabled` while `_tlsHandshakeDuration.Enabled` stays the same without the `Uncached` suffix since it does not need to be cached. I think that should be enough to avoid most slip ups.

Reply:
Removed public properties and added `CreateContext` method.